### PR TITLE
Fixes for Windows

### PR DESCRIPTION
**Pull request by Ace314159**
### Description

With this PR, MoveIt2 builds on Windows, and I'm able to use it for some basic planning. Here's a summary of all the changes:
1. Since there is no pkg-config on Windows, I had to find `FCL` and `Bullet` using find_package. Using vcpkg, the name for FCL is FCL instead of LIBFCL, so I also had to modify it to get it to work with that.
2. Because static variables are not exported by the global export I had to add manual visibility control to these libraries:
    - `moveit_collision_detection`
    - `moveit_collision_detection_fcl`
    - `moveit_collision_detection_bullet`
    - `moveit_kinematics_base`
    - `moveit_planning_scene`
    - `moveit_mesh_filter`
    - `moveit_planning_scene_monitor`
    - `moveit_warehouse`
    - `moveit_planning_pipeline`
    - `moveit_trajectory_execution_manager`
    - `moveit_move_group_interface`
    - `moveit_ros_robot_interaction`
    - `moveit_planning_scene_rviz_plugin`
3. The `zlib` component for boost is no longer required for `iostreams`
4. RDF loader uses `popen` and `pclose`, but the Windows equivalents are `_popen` and `_pclose`
5. MoveIt ROS warehouse connector uses `fork` and `kill`, but there's no equivalent `fork` and `kill` on Windows (as far as I'm aware), so I just disabled it on Windows and log an error when it's used
6. Windows adds some extra defines like `near`, `far`, and `max` that are used as variable names. Using `undef` fixes these errors. Also, `uint` doesn't work on MSVC, so I replaced that to `unsigned int`
7. `BenchmarkExecutor` does include `winsock2.h` on Windows, but it also needs to link with `ws2_32.lib` to build
8. In collision detection fcl and bullet, the tests are built with some compiler flags (`-Wno-deprecated-declarations`) that don't work on Windows, so I didn't add those flags on Windows.
9. I had to add a couple of includes (`deque` and `boost/bind.hpp`) in some places to get some things to compile
10. I changed the include from `pluginlib/class_list_macros.h` to `pluginlib/class_list_macros.hpp` in a couple of places to fix the deprecation warning. The warning would actually cause it to not compile because `#warning` isn't supported on MSVC and that causes a compile error

There are also some of MoveIt2's dependencies that have PRs that need to be merged in order for MoveIt2 to build and completely work on Windows:
- https://github.com/ros2/rviz/pull/718
- https://github.com/ros2/geometry2/pull/434
- https://github.com/ros-controls/ros2_control/pull/443
- https://github.com/ros-perception/vision_opencv/pull/390

I know #238 already exists with some fixes for windows, but it is out of date and doesn't have all of the fixes this PR has.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by davetcoleman** [comment]
Tagging @lilustga from Microsoft to review, original author of https://github.com/ros-planning/moveit2/pull/238

Also tagging MS Robotics lead @ooeygui

**Comment by lilustga** [comment]
Hi @davetcoleman, @ooeygui and I met with @Ace314159 recently to discuss this work and getting changes merged. I'll do a full review today.

File: moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h
Comment (lilustga) [inline moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h:160]: You may be overwriting this commit:
https://github.com/ros-planning/moveit2/commit/fb94e78da7d9a16fee53ccc34fbfb2c7af2af5d0

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (lilustga) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:581]: As mentioned above, you may be overwriting this commit:
https://github.com/ros-planning/moveit2/commit/fb94e78da7d9a16fee53ccc34fbfb2c7af2af5d0

File: moveit_ros/perception/mesh_filter/CMakeLists.txt
Comment (lilustga) [inline moveit_ros/perception/mesh_filter/CMakeLists.txt:18]: Instead of ${GLEW_LIBRARIES}, I have found it's more reliable portability wise to use GLEW::glew

File: moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp
Comment (lilustga) [inline moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp:169]: It looks like you are overwriting this commit:
https://github.com/ros-planning/moveit2/commit/bf6fc9afb86e9d1d795007e822af97ff63432721

File: moveit_core/robot_model/include/moveit/robot_model/joint_model_group.h
Comment (lilustga) [reply moveit_core/robot_model/include/moveit/robot_model/joint_model_group.h:139]: Ah I think you are overwriting this commit: https://github.com/ros-planning/moveit2/commit/68b2dc3af62a8c70e273f6be695b79d5a6372e0c

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_display.cpp
Comment (lilustga) [inline moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_display.cpp:210]: Overwrites this commit: https://github.com/ros-planning/moveit2/commit/eb2c8b97862e9acb760dd0c7882d06d04e4ecfae

File: moveit_ros/warehouse/warehouse/src/warehouse_connector.cpp
Comment (lilustga) [inline moveit_ros/warehouse/warehouse/src/warehouse_connector.cpp:44]: My approach here was to ifdef out the whole function so it does nothing but return false:
https://github.com/ms-iot/moveit2/commit/b56eba13e5fb7ab594f774b0b51cdd8f03dad5e2

I like adding the error message to make it clear what went wrong.

File: moveit_core/collision_detection/include/moveit/collision_detection/visibility_control.h
Comment (lilustga) [reply moveit_core/collision_detection/include/moveit/collision_detection/visibility_control.h:1]: for example:
https://github.com/ms-iot/moveit2/blob/master/moveit_core/collision_detection/include/moveit/collision_detection/visibility_control.hpp

File: moveit_core/collision_detection_bullet/CMakeLists.txt
Comment (Ace314159) [reply moveit_core/collision_detection_bullet/CMakeLists.txt:65]: If someone were using GCC using MinGW on Windows, they would still be able to use those flags. The intention was to only allow it for GCC and Clang. This pattern is used in existing code like [Rviz](https://github.com/ros2/rviz/blob/ros2/rviz_rendering_tests/CMakeLists.txt)

**Comment by Ace314159** [comment]
@lilustga I've made the changes. I'd love it if you could look at it again

File: moveit_core/collision_detection_bullet/CMakeLists.txt
Comment (lilustga) [reply moveit_core/collision_detection_bullet/CMakeLists.txt:65]: Ok, sounds reasonable

**Comment by lilustga** [comment]
Hi @Ace314159, the changes look good. I'm going to clone it today and test with `run_moveit_cpp`

**Comment by Ace314159** [comment]
Since not all the PRs for the dependencies have merged yet, you might have to use my forks for them. I've linked the PRs above, but here's them again:
- https://github.com/ros2/rviz/pull/718
- https://github.com/ros2/geometry2/pull/434
- https://github.com/ros-perception/vision_opencv/pull/390

Let me know if you have any issues

**Comment by lilustga** [comment]
Hi @Ace314159 I cloned and tried to build earlier today and hit a build error in moveit_core. Bullet couldn't be found. Looking into why not.

**Comment by Ace314159** [comment]
If you have pkg-config installed, Bullet will only be found using pkg-config. Vcpkg doens't provide a .pc file for bullet, so it won't be found. I wasn't sure how common this scenario was, so I hadn't tried supporting it. Also, with the .pc file vcpkg provides for FCL, there will be linking errors because the it uses the name FCL instead LIBFCL.

I'm not sure if using pkg-config is needed anymore for FCL and Bullet, so I got rid of it. If the CI fails, I'll add it back.

**Comment by Ace314159** [comment]
@lilustga pkg-config was needed for FCL but not Bullet. `find_package` will be used to find FCL before trying pkg-config. I've made those changes and the CI passes. Please try building again. It should work even if you have pkg-config installed

**Comment by lilustga** [comment]
@Ace314159 I hit a few more issues while building:
* Needed to change ${GLEW_LIBRARIES} to GLEW::glew in moveit2\moveit_ros\perception\CMakeLists.txt
* Found that I needed to add LIBCCD for moveit_core to build (ms-iot fork https://github.com/ros-planning/moveit2/blob/b56eba13e5fb7ab594f774b0b51cdd8f03dad5e2/moveit_core/CMakeLists.txt#L33)
* Had a linking failure about MD5 without OpenSSL had to revert commit 42efad9a53fe14ec48c237d10143259fdc33df1f

**Comment by Ace314159** [comment]
@lilustga For the MD5 issue, have you tried updating `warehouse_ros`? The linked PR talks about how building will fail unless that is updated. 

I didn't need the LIBCCD and GLEW changes to build. I have a [working CI run](https://github.com/Ace314159/ros2-foxy-windows-packages/runs/3026276614?check_suite_focus=true) that builds all of MoveIt2 and its dependencies. Can you try and see if running the steps in the workflow work for you? Or if there's a difference in our environments causing the issue?

**Comment by Ace314159** [comment]
@lilustga I was a bit hesitant about manually finding LIBCCD and linking with it because it is a dependency for FCL, not moveit_core, so I expected linking FCL to be enough. However, I was able to figure out why it wasn't being linked. It was because `${FCL_LIBRARIES}` is a list containing fcl and its dependencies, but `find_library` interprets the list as alternate names for the same single library. Looping through the list and using `find_library` on all of them fixed that.

Switching to `GLEW::glew` didn't affect my build, so I changed ${GLEW_LIBRARIES} to that as well.

However, I still think it would be useful to identify the differences between our environments. Here's a [link to my GitHub Actions workflow file](https://github.com/Ace314159/ros2-foxy-windows-packages/blob/6e583896788ecb4b520614a4c1c35b2800ce6fb4/.github/workflows/workspaces.yml) that didn't need these changes.

**Comment by lilustga** [comment]
@Ace314159 nice catch on the `${FCL_LIBRARIES}` bug. In your environment I don't see anything regarding the mongo dependency, are you also building warehouse_ros_mongo?

**Comment by Ace314159** [comment]
@lilustga I don't build `warehouse_ros_mongo`, but I can still build moveit2 and plan without it. I didn't build it because we weren't using it and @nkalupahana tried building it for macOS and experienced annoying issues that we didn't want to solve.

**Comment by nkalupahana** [comment]
Yeah, just to jump in here -- building the required mongo driver requires Python 2.6 and a very strange build tool (scons) that has moved way past its Python 2.6 days (as has everyone else). Thus, it's an absolute pain to get set up and build. We ran into serious issues getting this set up in any reasonable way, and we don't need it in our setup, so we've de-prioritized getting it working. If you've figured out a way to get this built in CI/similar, let us know :)

**Comment by traversaro** [comment]
> Because static variables are not exported by the global export I had to add manual visibility control to these libraries:

Just as a cross-link for anyone interested to be aware on this, in moveit1 the manual visibility headers was initially proposed in https://github.com/ros-planning/moveit/pull/2604, but in https://github.com/ros-planning/moveit/pull/2604#discussion_r623960973 it was requested to change them to use `GenerateExportHeader` CMake module.

**Comment by lilustga** [comment]
This was the same issue discussed here: https://github.com/ros-planning/moveit2/pull/286

It was ultimately the only thing that stopped the Windows fixes from going in back then. I support changing to the generated headers as requested in the interest of getting the changes merged.

@Ace314159 how do you feel about switching to the generated visibility headers? Link for reference: https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/

Also, I'm updating the foxy build on windows (2021-06-30) and this branch builds and runs the demo on Windows 👍.

**Comment by Ace314159** [comment]
Yes, I'm happy to make those changes in order to get these changed merged, and I'm glad you were able to run the demo!

@lilustga I'd also like your input on https://github.com/ros2/geometry2/pull/434. I'm not sure what the best way to proceed is, and if those changes don't get merged, we will have to add the `/permissive-` option when building.

**Comment by Ace314159** [comment]
@lilustga I've switched to using the generated headers. How does it look?

**Comment by lilustga** [comment]
@Ace314159 I hit some linker errors when building after the change. I made a PR to your fork with the changes I made to fix them.

**Comment by Ace314159** [comment]
@lilustga Sorry about that. I also made similar changes for the other classes since your PR didn't change all the classes.

**Comment by lilustga** [comment]
@traversaro @henningkayser Is there anything else you think should be changed before approval?

It looks like the workflows won't run until then.

**Comment by traversaro** [comment]
Fine for me (even if I am not someone that can trigger ci).

**Comment by lilustga** [comment]
@davetcoleman next steps before getting approval and running the workflows?

**Comment by wolfv** [comment]
it would be great to get this in -- we're working towards releasing moveit and other packages for galactic on Windows and macOS in the RoboStack project. https://robostack.github.io/galactic.html

RViz is already working, would be cool to get some MoveIt and Navigation2 working on Windows as well!

**Comment by vatanaksoytezer** [comment]
@lilustga @Ace314159 as we've discussed in the maintainers meeting we will have some people testing and approving this on their Windows setups. Would you mind pointing out instructions to get moveit2 compiled and running on Windows?

**Comment by lilustga** [comment]
I would recommend following these instructions:
http://wiki.ros.org/Installation/Windows

For step 5, run the following to get the new foxy pre-release which includes moveit2:
```
mkdir c:\opt\chocolatey
set ChocolateyInstall=c:\opt\chocolatey
choco source add -n=ros-win -s="https://aka.ms/ros/public" --priority=1
choco upgrade ros-foxy-desktop --pre -y --execution-timeout=0
```

All the dependencies for building moveit2 should be included at this point. You can build by cloning the moveit2 into a new workspace and running `colcon build` in a ROS command prompt (as described in the Windows Installation instructions).

Let me know if you hit any issues!

**Comment by mamoll** [comment]
@lilustga I am trying to build this on my Windows VM. I got pretty far, but there are two issues:
- [Minor] I installed MongoDB via `choco`, but it doesn't include header files, so I have to ignore the `warehouse_ros_mongo` package: 
```
colcon build --packages-skip warehouse_ros_mongo --event-handlers desktop_notification- status- --cmake-args -DCMAKE_BUILD_TYPE=Release
```
- [Fatal] It's looking for non-existing debug versions of some Boost libraries:
```
c:\dev_moveit2\build\moveit_planners_ompl>cmake --build .
Microsoft (R) Build Engine version 16.7.0+b89cb5fde for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Auto build dll exports
LINK : warning LNK4044: unrecognized option '/openmp'; ignored [C:\dev_moveit2\build\moveit_planners_ompl\om
pl_interface\moveit_ompl_interface.vcxproj]
LINK : fatal error LNK1104: cannot open file 'C:\opt\ros\foxy\x64\debug\lib\boost_serialization-vc140-mt-gd.
lib' [C:\dev_moveit2\build\moveit_planners_ompl\ompl_interface\moveit_ompl_interface.vcxproj]
  gtest.vcxproj -> C:\dev_moveit2\build\moveit_planners_ompl\gtest\Debug\gtest.lib
  gtest_main.vcxproj -> C:\dev_moveit2\build\moveit_planners_ompl\gtest\Debug\gtest_main.lib
```
The non-debug versions exist `C:\opt\ros\foxy\x64\lib`. There's probably a simple fix for this, but I am not too familiar with Windows development.

**Comment by lilustga** [comment]
@mamoll The mongo DB isn't working on Windows or Mac right now, you can see some conversation above in the thread about warehouse_ros_mongo.

For the second issue, are you getting the same error when you call colcon build (instead of cmake directly)?

Which version of Visual Studio do you have installed? Do you have the "Desktop development with C++" workload included?

**Comment by mamoll** [comment]
I realized that I should have typed `cmake --build . --target Release`. Then the code _almost_ compiles. I just needed to make this one-line change:
```
--- a/moveit_planners/ompl/ompl_interface/scripts/generate_state_database.cpp
+++ b/moveit_planners/ompl/ompl_interface/scripts/generate_state_database.cpp
@@ -56,7 +56,7 @@ static const std::string ROBOT_DESCRIPTION = "robot_description";
 static const std::string CONSTRAINT_PARAMETER = "constraints";

 static bool get_uint_parameter_or(const rclcpp::Node::SharedPtr& node, const std::string& param_name,
-                                  size_t&& result_value, const size_t default_value)
+                                  size_t result_value, size_t default_value)
 {
   int param_value;
   if (node->get_parameter(param_name, param_value))
```
It makes no sense to make POD types universal refs., so this seems like a harmless change. (The const change isn't strictly necessary, but also doesn't make sense in the original code.)

**Review by mamoll — APPROVED** [review]
This looks good to me. I had to change one function to get it to compile. Not sure what is different about my setup. Let's get this merged and let's add some Windows configs to the CI build matrix! Nice work, @Ace314159 and thanks for the assist, @lilustga.

**Comment by tylerjw** [comment]
> It seems to fail when trying to install the upstream dependencies. I don't think that's related to the changes.

This is affecting all of our PRs right now.  The issue seems to be with the official build farm.  We are tracking the issue here: https://github.com/osrf/infrastructure/issues/5

I'll retrigger the CI for this job (and the others) when that is resolved.

**Comment by wyattrees** [comment]
Tested this on my Windows system and I'm running into this error from `moveit_ros_perception` when building:

```
semantic_world.obj : error LNK2019: unresolved external symbol "void __cdecl cv::findContours(class cv::_InputArray const &,class cv::_OutputArray const &,class cv::_OutputArray const &,int,int,class cv::Point_<int>)" (?findContours@cv@@YAXAEBV_InputArray@1@AEBV_OutputArray@1@1HHV?$Point_@H@1@@Z) referenced in function "public: class std::vector<struct geometry_msgs::msg::PoseStamped_<class std::allocator<void> >,class std::allocator<struct geometry_msgs::msg::PoseStamped_<class std::allocator<void> > > > __cdecl moveit::semantic_world::SemanticWorld::generatePlacePoses(struct object_recognition_msgs::msg::Table_<class std::allocator<void> > const &,double,double,double,unsigned int,double)const " (?generatePlacePoses@SemanticWorld@semantic_world@moveit@@QEBA?AV?$vector@U?$PoseStamped_@V?$allocator@X@std@@@msg@geometry_msgs@@V?$allocator@U?$PoseStamped_@V?$allocator@X@std@@@msg@geometry_msgs@@@std@@@std@@AEBU?$Table_@V?$allocator@X@std@@@msg@object_recognition_msgs@@NNNIN@Z) [C:\dev_ws\build\moveit_ros_perception\semantic_world\moveit_semantic_world.vcxproj]
C:\dev_ws\build\moveit_ros_perception\semantic_world\Release\moveit_semantic_world.dll : fatal error LNK1120: 1 unresolved externals [C:\dev_ws\build\moveit_ros_perception\semantic_world\moveit_semantic_world.vcxproj]
```

Note: the package silently failed until I included `--event-handlers console_direct+` along with `colcon build`

I have OpenCV installed, and it is found during the build process:

`-- Found OpenCV 3.4.6 in C:/opencv/x64/vc16/lib`

any suggestions on how to fix this?

**Comment by lilustga** [comment]
@wyattrees I haven't been able to repro this, I'm building this branch successfully with `colcon build --event-handlers console_direct+`.

Do you see linker errors for all the OpenCV symbols in semantic_world (`cv::Point`, `cv::Mat`, etc.) or just for `cv::findContours`?

**Comment by Ace314159** [comment]
@wyattrees I'd also recommend using [vcpkg](https://github.com/microsoft/vcpkg) for dependency management, which is much simpler to use and what I have tested with. I also have a GitHub Action that successfully builds MoveIt2 and its dependencies using vcpkg [here](https://github.com/Ace314159/ros2-foxy-windows-packages/runs/3158465663?check_suite_focus=true) if you'd like to see how to use it.

**Comment by lilustga** [comment]
@Ace314159 with this change https://github.com/ros-planning/moveit2/pull/576 tests should pass now. If you rebase or merge with main again I think the PR will be ready to merge.

**Comment by lilustga** [comment]
@tylerjw @henningkayser Now that this is green, how do you feel about merging?

**Review by tylerjw — APPROVED** [review]
I read through these changes and don't see anything wrong.  I think we should merge this so the windows team can use moveit on foxy with our latest fixes.  This will also the be the last thing we are waiting on before making a foxy branch.

**Comment by wyattrees** [comment]
@lilustga @Ace314159 I'm only seeing the linker error for findContours. I tried using vcpkg to install the dependencies but still getting the same. I think this is probably an issue with something else I'm doing wrong in setup. Anyway, thanks for your work on this!

**Pull request merged**
